### PR TITLE
Allowed ignoring of scalar time coords in `multi_model_statistics`

### LINKED ISSUE #1960: `ignore_scalar_coords` option of multi-model statistics does not work with scalar time coordinate
The option [`ignore_scalar_coords`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html?highlight=ignore_scalar_coords#multi-model-statistics) does not work properly for input cubes with different scalar time coordinates (i.e., we get an error).

The reason is that the alignment of the time dimensions happens before the removal of scalar coordinates by `ignore_scalar_coords`. This alignment does not work with differing scalar time coordinates since it uses [`iris.cube.Cube.interpolate`](https://github.com/ESMValGroup/ESMValCore/blob/5eb8f91447695e63a2955bef0dde8dc2fd46ea07/esmvalcore/preprocessor/_multimodel.py#L199), which makes no sense to use on scalar coordinates and rightfully fails with a

```
ValueError: Coordinates repeat a data dimension - the interpolation would be over-specified.
```

The simple solution to this is to remove the scalar coordinates before the time alignment.

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR makes the usage of the option `ignore_scalar_coords=True` in `multi_model_statistics` work for scalar time coordinates.

Closes #1960

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_multimodel.py [hunk 1]
```diff
@@ -198,9 +198,17 @@ def _map_to_new_time(cube, time_points):
     try:
         new_cube = cube.interpolate(sample_points, scheme)
     except Exception as excinfo:
+        additional_info = ""
+        if cube.coords('time', dimensions=()):
+            additional_info = (
+                " Note: this alignment does not work for scalar time "
+                "coordinates. To ignore all scalar coordinates in the input "
+                "data, use the option `ignore_scalar_coords=True`."
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_multimodel.py:206]:
```suggestion
                "data, use the preprocessor option `ignore_scalar_coords=True`."
```

File: esmvalcore/preprocessor/_multimodel.py [hunk 2]
```diff
@@ -356,16 +364,13 @@ def _equalise_coordinate_metadata(cubes, ignore_scalar_coords=False):
...
+        # at this point anymore.
         scalar_coords_to_always_remove = ['p0', 'ptop']
         for scalar_coord in cube.coords(dimensions=()):
-            remove_coord = (
-                ignore_scalar_coords or
-                scalar_coord.var_name in scalar_coords_to_always_remove
-            )
-            if remove_coord:
+            if scalar_coord.var_name in scalar_coords_to_always_remove:
                 cube.remove_coord(scalar_coord)
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_multimodel.py:374]: probably best to toss a warning in debug, I think

File: esmvalcore/preprocessor/_multimodel.py [hunk 3]
```diff
@@ -570,9 +574,16 @@ def _multicube_statistics(cubes, statistics, span, ignore_scalar_coords=False):
     # Avoid modifying inputs
     copied_cubes = [cube.copy() for cube in cubes]
 
+    # Remove scalar coordinates in input cubes if desired to ignore them when
+    # merging
+    if ignore_scalar_coords:
+        for cube in copied_cubes:
+            for scalar_coord in cube.coords(dimensions=()):
+                cube.remove_coord(scalar_coord)
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_multimodel.py:582]: here too - a warning or simple debug message

**Review by valeriupredoi — APPROVED** [review]
looking good, cheers Manu! Just a couple minor points from me :beer:

File: tests/sample_data/multimodel_statistics/test_multimodel.py [hunk 4]
```diff
@@ -396,3 +421,70 @@ def test_multimodel_0d_different_time_dimensions(timeseries_cubes_month):
...
         multimodel_test(cubes, span=span, statistic='mean')
+
+
+@pytest.mark.use_sample_data
+@pytest.mark.parametrize('span', SPAN_PARAMS)
+def test_multimodel_diff_scalar_time_ignore(timeseries_cubes_month, span):
+    """Ignore different scalar time dimensions.
+
+    Also remove air_pressure dimensions since this slightly differs across
+    cubes. See https://github.com/ESMValGroup/ESMValCore/issues/956.
```
Comment (bouweandela) [inline tests/sample_data/multimodel_statistics/test_multimodel.py:432]: As far as I can see from the code and from running a test, the `air_pressure` coordinate is not removed, so I'm a bit confused by this comment. Could you explain?

Comment (schlunma) [reply tests/sample_data/multimodel_statistics/test_multimodel.py:432]: If you use `timeseries_cubes_month` without any changes, the test will (rightfully) [fail](https://github.com/ESMValGroup/ESMValCore/blob/dc09915a48a000b011e1ce14e396541d0d70eaa6/tests/sample_data/multimodel_statistics/test_multimodel.py#L201-L216) due to different `air_pressure` coordinates:

```py
cube 0 [100000.  92500.]
cube 1 [100000.  92500.]
...
cube 12 [100000.00000001  92500.00000001]
cube 13 [100000.00000001  92500.00000001]
cube 14 [100000.  92500.]
...

```

I am not entirely sure why that happens, but it's probably described in https://github.com/ESMValGroup/ESMValCore/issues/956.

Thus, in this test (and many others), the `air_pressure` coordinate is collapsed by using `cube[0, 0]`. Note that different coordinate `points` are fine for scalar coordinates, just not different coordinate metadata.

I will adapt the comment to make it clear that the coordinate is collapsed, not removed.

Comment (bouweandela) [reply tests/sample_data/multimodel_statistics/test_multimodel.py:432]: Thanks! Now that we have the `esmvalcore.dataset.Dataset` class, we could use that to load the data and that should avoid the issue altogether. But that's something for another pull request.

**Pull request merged**